File: Leetcode/Sort Integers by The Power Value/Sort Integers by The Power Value.java
/*
The power of an integer x is defined as the number of steps needed to transform x into 1 using the following steps:

if x is even then x = x / 2
if x is odd then x = 3 * x + 1
For example, the power of x = 3 is 7 because 3 needs 7 steps to become 1 (3 --> 10 --> 5 --> 16 --> 8 --> 4 --> 2 --> 1).

Given three integers lo, hi and k. The task is to sort all integers in the interval [lo, hi] by the power value in ascending order, if two or more integers have the same power value sort them by ascending order.

Return the k-th integer in the range [lo, hi] sorted by the power value.

Notice that for any integer x (lo <= x <= hi) it is guaranteed that x will transform into 1 using these steps and that the power of x is will fit in 32 bit signed integer.


*/


class Solution {
    public int getKth(int lo, int hi, int k) {
        int len = hi - lo + 1;
        // 2d arr - [0] is x, [1] is x's power value
        int[][] arr = new int[len][2];
        for (int i = 0; i < len; i++) {
            arr[i][0] = lo + i;
            arr[i][1] = transform(lo + i);
        }
        
        // sort by index [1] which is x's power value
        Arrays.sort(arr, (a,b) -> Integer.compare(a[1], b[1]));

        // Here k-1 because of 0 indexing
        return arr[k-1][0];
    }
    
    public int transform(int x) {
        int temp = x;
        int count = 0;
        while (temp != 1) {
            temp = reduce(temp);
            count++;
        }
        // return power of x    
        return count;
    }
    
    // if x is even then x = x / 2
    // if x is odd then x = 3 * x + 1
    public int reduce(int x) {
        if (x % 2 == 0) x /= 2;
        else x = 3 * x + 1;
        return x;        
    }
}

File: Leetcode/Sort Integers by The Power Value/Sort Integers by The Power Value.py
#Problem:https://leetcode-cn.com/problems/sort-integers-by-the-power-value/description/
class Solution:
    def getKth(self, lo: int, hi: int, k: int) -> int:q
        d = {}
        for i in range(10):
            d[2**i] = i
        #calculate power value of number 
        def helper(num):
            if num in d:
                return d[num]
            if num % 2:
                ans = helper(num*3+1)+1
            else:
                ans = helper(num//2)+1
            d[num] = ans
            return ans
        #list of elements between lo to hi
        nums = list(range(lo,hi+1))
        #list of all power value of elements 
        ans = [[helper(n),n] for n in nums] 
        #sort the list in ascending order of power value 
        ans.sort(key = lambda x:x[0])
        return ans[k-1][1]
